File: android/src/main/java/ru/inaliv/appmetrica/Converter.java
package ru.inaliv.appmetrica;

import android.location.Location;
import android.os.Build;

import io.appmetrica.analytics.AppMetricaConfig;
import io.appmetrica.analytics.ecommerce.ECommerceAmount;
import io.appmetrica.analytics.ecommerce.ECommerceCartItem;
import io.appmetrica.analytics.ecommerce.ECommerceOrder;
import io.appmetrica.analytics.ecommerce.ECommercePrice;
import io.appmetrica.analytics.ecommerce.ECommerceProduct;
import io.appmetrica.analytics.ecommerce.ECommerceReferrer;
import io.appmetrica.analytics.ecommerce.ECommerceScreen;
import io.appmetrica.analytics.profile.Attribute;
import io.appmetrica.analytics.profile.GenderAttribute;
import io.appmetrica.analytics.profile.UserProfile;
import io.appmetrica.analytics.profile.UserProfileUpdate;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;


public class Converter {
    /**
     * Сконвертирует конфигурацию JSONObject в объект для App Метрики
     *
     * @param config
     * @return
     */
    public static AppMetricaConfig toConfig(final JSONObject config) throws JSONException {
        final String apiKey = config.getString("apiKey");
        final AppMetricaConfig.Builder builder = AppMetricaConfig.newConfigBuilder(apiKey);

        if (config.has("handleFirstActivationAsUpdate")) {
            builder.handleFirstActivationAsUpdate(config.getBoolean("handleFirstActivationAsUpdate"));
        }
        if (config.has("locationTracking")) {
            builder.withLocationTracking(config.getBoolean("locationTracking"));
        }
        if (config.has("sessionTimeout")) {
            builder.withSessionTimeout(config.getInt("sessionTimeout"));
        }
        if (config.has("crashReporting")) {
            builder.withCrashReporting(config.getBoolean("crashReporting"));
        }
        if (config.has("nativeCrashReporting")) {
            builder.withNativeCrashReporting(config.getBoolean("nativeCrashReporting"));
        }
        if (config.has("appVersion")) {
            builder.withAppVersion(config.getString("appVersion"));
        }
        if (config.optBoolean("logs", false)) {
            builder.withLogs();
        }
        if (config.has("location")) {
            final Location location = Converter.toLocation(config.getJSONObject("location"));
            builder.withLocation(location);
        }

        return builder.build();
    }

    /**
     * Конвертирует конфигурацию JSONObject в объект Location для App Метрики
     *
     * @param location
     * @return
     * @throws JSONException
     */
    public static Location toLocation(final JSONObject location) throws JSONException {
        final Location yamLocation = new Location("Custom");

        if (location.has("latitude")) {
            yamLocation.setLatitude(location.getDouble("latitude"));
        }
        if (location.has("longitude")) {
            yamLocation.setLongitude(location.getDouble("longitude"));
        }
        if (location.has("altitude")) {
            yamLocation.setAltitude(location.getDouble("altitude"));
        }

        if (location.has("vAccuracy")) {
            yamLocation.setAccuracy((float) location.getDouble("vAccuracy"));
        }
        else if (location.has("accuracy")) {
            yamLocation.setAccuracy((float) location.getDouble("accuracy"));
        }

        if (location.has("hAccuracy") && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            yamLocation.setVerticalAccuracyMeters((float) location.getDouble("hAccuracy"));
        }

        if (location.has("course")) {
            yamLocation.setBearing((float) location.getDouble("course"));
        }
        if (location.has("speed")) {
            yamLocation.setSpeed((float) location.getDouble("speed"));
        }
        if (location.has("timestamp")) {
            yamLocation.setTime(location.getLong("timestamp"));
        }

//        yamLocation.setVerticalAccuracyMeters(); // TODO:


        return yamLocation;
    }

    /**
     * From:
     * {
     *     "sku": "779213",              // [!] Обязательный
     *     "name": "Продукт творожный «Даниссимо» 5.9%, 130 г.",
     *     "actualPrice": { ... },      // Смотри структуру toECommercePrice()
     *     "originalPrice": { ... },    // Смотри структуру toECommercePrice()
     *     "categoriesPath": ["Продукты", "Молочные продукты", "Йогурты"],
     *     "promocodes": ["BT79IYX", "UT5412EP"],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     *
     * @param product
     * @return
     * @throws JSONException
     */
    public static ECommerceProduct toECommerceProduct(final JSONObject product) throws JSONException {
        ECommerceProduct yamProduct = new ECommerceProduct(product.getString("sku"));

        if (product.has("name")) {
            yamProduct.setName(product.getString("name"));
        }

        if (product.has("actualPrice")) {
            yamProduct.setActualPrice(
                toECommercePrice(
                    product.getJSONObject("actualPrice")
                )
            );
        }

        if (product.has("originalPrice")) {
            yamProduct.setOriginalPrice(
                toECommercePrice(
                    product.getJSONObject("originalPrice")
                )
            );
        }

        if (product.has("categoriesPath")) {
            yamProduct.setCategoriesPath(
                toStringList(
                    product.getJSONArray("categoriesPath")
                )
            );
        }

        if (product.has("promocodes")) {
            yamProduct.setPromocodes(
                toStringList(
                    product.getJSONArray("promocodes")
                )
            );
        }

        if (product.has("payload")) {
            yamProduct.setPayload(
                toHashMapPayload(
                    product.getJSONObject("payload")
                )
            );
        }

        return yamProduct;
    }

    /**
     * From:
     * {
     *     "name": "ProductCardActivity",
     *     "searchQuery": "даниссимо кленовый сироп",
     *     "categoriesPath": ["Акции", "Красная цена"],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     *
     * NOTE: В SDK для iOS "categoriesPath" называется "categoryComponents"
     *
     * @param screen
     * @return
     * @throws JSONException
     */
    public static ECommerceScreen toECommerceScreen(final JSONObject screen) throws JSONException {
        ECommerceScreen yamScreen = new ECommerceScreen();

        if (screen.has("name")) {
            yamScreen.setName(screen.getString("name"));
        }

        if (screen.has("searchQuery")) {
            yamScreen.setSearchQuery(screen.getString("searchQuery"));
        }

        if (screen.has("categoriesPath")) {
            yamScreen.setCategoriesPath(toStringList(screen.getJSONArray("categoriesPath")));
        }

        if (screen.has("payload")) {
            yamScreen.setPayload(toHashMapPayload(screen.getJSONObject("payload")));
        }

        return yamScreen;
    }

    /**
     * From:
     * {
     *     "product": { ... }, // [!] Обязательный. Смотри структуру toECommerceProduct()
     *     "revenue": { ... }, // [!] Обязательный. Получаемый доход. Смотри структуру toECommercePrice()
     *     "quantity": 1.0,    // [!] Обязательный.
     *     "referrer": { ... } // Смотри структуру toECommerceReferrer()
     * }
     *
     * @param item
     * @return
     * @throws JSONException
     */
    public static ECommerceCartItem toECommerceCartItem(final JSONObject item) throws JSONException {
        ECommerceProduct yamProduct = toECommerceProduct(item.getJSONObject("product"));
        ECommercePrice yamRevenue = toECommercePrice(item.getJSONObject("revenue"));
        double quantity = (double)item.getDouble("quantity");

        ECommerceCartItem yamCartItem = new ECommerceCartItem(yamProduct, yamRevenue, quantity);

        if (item.has("referrer")) {
            yamCartItem.setReferrer(
                toECommerceReferrer(
                    item.getJSONObject("referrer")
                )
            );
        }

        return yamCartItem;
    }

    /**
     * From:
     * {
     *     "identifier": "88528768", // [!] Обязательный.
     *     "cartItems": [            // [!] Обязательный.
     *          { ... },             // Смотри структуру toECommerceCartItem()
     *          ...
     *     ],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     *
     * @param order
     * @return
     * @throws JSONException
     */
    public static ECommerceOrder toECommerceOrder(final JSONObject order) throws JSONException {
        String identifier = order.getString("identifier");
        List<ECommerceCartItem> yamCartItems = new ArrayList<>();

        JSONArray items = order.getJSONArray("cartItems");
        for (int i = 0; i < items.length(); ++i) {
            ECommerceCartItem yamItem = toECommerceCartItem(
                    items.getJSONObject(i)
            );

            yamCartItems.add(yamItem);
        }

        ECommerceOrder yamOrder = new ECommerceOrder(identifier, yamCartItems);

        if (order.has("payload")) {
            yamOrder.setPayload(
                toHashMapPayload(
                    order.getJSONObject("payload")
                )
            );
        }

        return yamOrder;
    }

    /**
     * From:
     * {
     *     "type": "button",
     *     "identifier": "76890",
     *     "screen": { ... }      // Смотри структуру toECommerceScreen
     * }
     *
     * @param referrer
     * @return
     * @throws JSONException
     */
    public static ECommerceReferrer toECommerceReferrer(final JSONObject referrer) throws JSONException {
        ECommerceReferrer yamReferrer = new ECommerceReferrer();

        if (referrer.has("type")) {
            yamReferrer.setType(referrer.getString("type"));
        }

        if (referrer.has("identifier")) {
            yamReferrer.setIdentifier(referrer.getString("identifier"));
        }

        if (referrer.has("screen")) {
            yamReferrer.setScreen(
                toECommerceScreen(
                    referrer.getJSONObject("screen")
                )
            );
        }

        return yamReferrer;
    }

    /**
     * From:
     * {
     *     "fiat": [4.53, "USD"],      // [!] Обязательный
     *     "internalComponents": [
     *          [30_570_000, "wood"],
     *          [26.89, "iron"],
     *          [5.1, "gold"]
     *     ]
     * }
     *
     * @param price
     * @return
     * @throws JSONException
     */
    public static ECommercePrice toECommercePrice(final JSONObject price) throws JSONException {
        ECommercePrice yamPrice = new ECommercePrice(
            toECommerceAmount(
                price.getJSONArray("fiat")
            )
        );

        if (price.has("internalComponents")) {
            JSONArray internalComponents = price.getJSONArray("internalComponents");
            List<ECommerceAmount> yamInternalComponents = new ArrayList<>();

            for (int i = 0; i < internalComponents.length(); ++i) {
                yamInternalComponents.add(
                    toECommerceAmount(
                        internalComponents.getJSONArray(i)
                    )
                );
            }

            yamPrice.setInternalComponents(yamInternalComponents);
        }

        return yamPrice;
    }

    /**
     * From:
     * [10.5, "USD"]
     *
     * @param jsonArray
     * @return
     * @throws JSONException
     */
    public static ECommerceAmount toECommerceAmount(JSONArray jsonArray) throws JSONException {
        return new ECommerceAmount(jsonArray.getDouble(0), jsonArray.getString(1));
    }

    /**
     * Используется для дополнительных значений в:
     * - toECommerceScreen
     * - toECommerceProduct
     *
     * @param payload
     * @return
     */
    public static Map<String, String> toHashMapPayload(final JSONObject payload) throws JSONException {
        Map<String, String> yamPayload = new HashMap<>();
        Iterator<String> payloadIter = payload.keys();

        while (payloadIter.hasNext()) {
            String key = payloadIter.next();
            yamPayload.put(key, payload.getString(key));
        }

        return yamPayload;
    }

    /**
     * Преобразует JSONArray в строковый список, используется в
     * - toECommerceScreen, для категорий
     * - toECommerceProduct, для категорий и промокодов
     *
     * @param jsonArray
     * @return
     * @throws JSONException
     */
    private static List<String> toStringList(JSONArray jsonArray) throws JSONException {
        List<String> list = new ArrayList<>();
        for (int i = 0; i < jsonArray.length(); ++i) {
            list.add(jsonArray.get(i).toString());
        }

        return list;
    }

    /**
     * From:
     * {
     *   "name": "Ivan Nalivayko",
     *   "gender": "male",
     *   "notificationEnabled": false,
     *   "birthDate": { // OR: { "age": 20 }
     *     "year": 2001,
     *     "month": 1,
     *      "day": 1
     *   }
     * }
     *
     * @param user
     * @return
     * @throws JSONException
     */
    public static UserProfile toUserProfile(final JSONObject user) throws JSONException {
        UserProfile.Builder yamProfileBuilder = UserProfile.newBuilder();

        if (user.has("name")) {
            yamProfileBuilder.apply(Attribute.name().withValue(
                user.getString("name")
            ));
        }

        if (user.has("gender")) {
            yamProfileBuilder.apply(Attribute.gender().withValue(
                toGenderType(user.getString("gender"))
            ));
        }

        if (user.has("notificationEnabled")) {
            yamProfileBuilder.apply(Attribute.notificationsEnabled().withValue(
                user.getBoolean("notificationEnabled")
            ));
        }

        if (user.has("birthDate")) {
            JSONObject birthDate = user.getJSONObject("birthDate");
            UserProfileUpdate yamBirthDate = null;

            if (birthDate.has("year")) {
                int year = birthDate.getInt("year");

                if (birthDate.has("month")) {
                    int month = birthDate.getInt("month");

                    if (birthDate.has("day")) {
                        int day = birthDate.getInt("day");

                        yamBirthDate = Attribute.birthDate().withBirthDate(year, month, day);
                    } else {
                        yamBirthDate = Attribute.birthDate().withBirthDate(year, month);
                    }
                } else {
                    yamBirthDate = Attribute.birthDate().withBirthDate(year);
                }
            } else if (birthDate.has("age")) {
                int age = birthDate.getInt("age");

                yamBirthDate = Attribute.birthDate().withAge(age);
            }

            if (yamBirthDate != null) {
                yamProfileBuilder.apply(yamBirthDate);
            }
        }

        return yamProfileBuilder.build();
    }

    private static GenderAttribute.Gender toGenderType(final String gender) {
        return switch (gender.toLowerCase(Locale.ROOT)) {
            case "female" -> GenderAttribute.Gender.FEMALE;
            case "male" -> GenderAttribute.Gender.MALE;
            default -> GenderAttribute.Gender.OTHER;
        };
    }
}

File: ios/Plugin/Converter.swift
//
//  Converter.swift
//  CapacitorAppmetricaPlugin
//
//  Created by Nalivayko Ivan on 22.07.2022.
//

import Foundation
import YandexMobileMetrica
import CoreLocation

class Converter {
    enum ValidationError: Error {
        case apiKeyNotDefined
        case incorrectAmount
        case incorrectProductSku
        case incorrectProductItemQty
        case incorrectOrderId
        case incorrectOrderCartItems
        case unknown
    }
    
    /*
     * Конвертирует JSObject конфигурацию в объект конфигурации для AppMetrika
     */
    static func toConfig(config: [AnyHashable: Any]) throws -> YMMYandexMetricaConfiguration {
        guard let apiKey = config["apiKey"] as? String else {
            throw ValidationError.apiKeyNotDefined
        }
        
        let yamConfig = YMMYandexMetricaConfiguration.init(apiKey: apiKey)!
        
        if let handleFirstActivationAsUpdate = config["handleFirstActivationAsUpdate"] as? Bool {
            yamConfig.handleFirstActivationAsUpdate = handleFirstActivationAsUpdate
        }
        
        if let locationTracking = config["locationTracking"] as? Bool {
            yamConfig.locationTracking = locationTracking
        }
        
        if let sessionTimeout = config["sessionTimeout"] as? UInt {
            yamConfig.sessionTimeout = sessionTimeout
        }
        
        if let crashReporting = config["crashReporting"] as? Bool {
            yamConfig.crashReporting = crashReporting
        }
        
        if let appVersion = config["appVersion"] as? String {
            yamConfig.appVersion = appVersion
        }
        
        if let logs = config["logs"] as? Bool {
            yamConfig.logs = logs
        }
        
        if let location = config["location"] as? [AnyHashable: Any] {
            yamConfig.location = self.toLocation(location: location)
        }
        
        return yamConfig
    }
    
    /*
     * Конвертирует JSObject в объект местоположения
     */
    static func toLocation(location: [AnyHashable: Any]) -> CLLocation {
        let latitude        = location["latitude"]  as? Double   ?? 0.0
        let longitude       = location["longitude"] as? Double   ?? 0.0
        let altitude        = location["altitude"]  as? Double   ?? 0.0
        let hAccuracy       = location["hAccuracy"] as? Double   ?? location["accuracy"] as? Double ?? 0.0
        let vAccuracy       = location["vAccuracy"] as? Double   ?? -1
        let timestampNumber = location["timestamp"] as? NSNumber ?? -1
        let course          = location["course"]    as? Double   ?? 0.0
        let speed           = location["speed"]     as? Double   ?? 0.0
        
        let locationDate = Date(timeIntervalSince1970: timestampNumber.doubleValue)
        let coordinate = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        
        let yamLocation = CLLocation(
            coordinate:         coordinate,
            altitude:           altitude,
            horizontalAccuracy: hAccuracy,
            verticalAccuracy:   vAccuracy,
            course:             course,
            speed:              speed,
            timestamp:          locationDate
        )
        
        return yamLocation
    }
    
    /*
     * From:
     * {
     *     "name": "ProductCardActivity",
     *     "searchQuery": "даниссимо кленовый сироп",
     *     "categoriesPath": ["Акции", "Красная цена"],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     *
     * NOTE: В SDK для iOS "сategoriesPath" называется "categoryComponents"
     */
    static func toECommerceScreen(screen: [AnyHashable: Any]) -> YMMECommerceScreen {
        let yamScreen = YMMECommerceScreen(
            name:               screen["name"] as? String,
            categoryComponents: self.toStringArray(screen["categoriesPath"]),
            searchQuery:        screen["searchQuery"] as? String,
            payload:            screen["payload"] as? [String:String]
        )
        
        return yamScreen
    }
    
    /*
     * From:
     * {
     *     "sku": "779213",              // [!] Обязательный
     *     "name": "Продукт творожный «Даниссимо» 5.9%, 130 г.",
     *     "actualPrice": { ... },       // Смотри структуру toECommercePrice()
     *     "originalPrice": { ... },     // Смотри структуру toECommercePrice()
     *     "categoriesPath": ["Продукты", "Молочные продукты", "Йогурты"],
     *     "promocodes": ["BT79IYX", "UT5412EP"],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     */
    static func toECommerceProduct(product: [AnyHashable: Any]) throws -> YMMECommerceProduct {
        guard let sku = product["sku"] as? String else {
            throw ValidationError.incorrectProductSku
        }
        
        var actualPrice: YMMECommercePrice? = nil
        var originalPrice: YMMECommercePrice? = nil
        
        if product.index(forKey: "actualPrice") != nil {
            actualPrice = try self.toECommercePrice(price: product["actualPrice"] as? [AnyHashable: Any] ?? [:])
        }
        
        if product.index(forKey: "originalPrice") != nil {
            originalPrice = try self.toECommercePrice(price: product["originalPrice"] as? [AnyHashable: Any] ?? [:])
        }
        
        let yamProduct = YMMECommerceProduct(
            sku:                sku,
            name:               product["name"] as? String,
            categoryComponents: self.toStringArray(product["categoriesPath"]),
            payload:            product["payload"] as? [String:String],
            actualPrice:        actualPrice,
            originalPrice:      originalPrice,
            promoCodes:         self.toStringArray(product["promocodes"])
        )
        
        return yamProduct
    }
    
    /*
     * From:
     * {
     *     "product": { ... },  // [!] Обязательный. Смотри структуру toECommerceProduct()
     *     "revenue": { ... },  // [!] Обязательный. Получаемый доход. Смотри структуру toECommercePrice()
     *     "quantity": 1.0,     // [!] Обязательный.
     *     "referrer": { ... }  // Смотри структуру toECommerceReferrer()
     * }
     */
    static func toECommerceCartItem(item: [AnyHashable: Any]) throws -> YMMECommerceCartItem {
        let yamProduct = try self.toECommerceProduct(product: item["product"] as? [AnyHashable: Any] ?? [:])
        let yamRevenue = try self.toECommercePrice(price: item["revenue"] as? [AnyHashable: Any] ?? [:])
        
        guard let quantity = item["quantity"] as? NSNumber else {
            throw ValidationError.incorrectProductItemQty
        }
        
        var yamReferrer: YMMECommerceReferrer? = nil
        if item.index(forKey: "referrer") != nil {
            yamReferrer = self.toECommerceReferrer(referrer: item["referrer"] as? [AnyHashable: Any] ?? [:])
        }
        
        let yamCartItem = YMMECommerceCartItem(
            product:    yamProduct,
            quantity:   NSDecimalNumber(value: quantity.doubleValue),
            revenue:    yamRevenue,
            referrer:   yamReferrer
        )
        
        return yamCartItem
    }
    
    /*
     * From:
     * {
     *     "identifier": "88528768",     [!] Обязательный.
     *     "cartItems": [                [!] Обязательный.
     *          { ... },                 Смотри структуру toECommerceCartItem()
     *          ...
     *     ],
     *     "payload": {
     *         "ключ": "текстовое значение",
     *         ...
     *     }
     * }
     */
    static func toECommerceOrder(order: [AnyHashable: Any]) throws -> YMMECommerceOrder {
        guard let identifier = order["identifier"] as? String else {
            throw ValidationError.incorrectOrderId
        }
        
        guard let items = order["cartItems"] as? [ [AnyHashable: Any] ] else {
            throw ValidationError.incorrectOrderCartItems
        }
        
        let yamCartItems = try items.map { item in
            return try self.toECommerceCartItem(item: item)
        }
        
        let yamOrder = YMMECommerceOrder(
            identifier: identifier,
            cartItems:  yamCartItems,
            payload:    order["payload"] as? [String: String]
        )
        
        return yamOrder
    }
    
    /*
     * From:
     * {
     *     "type": "button",
     *     "identifier": "76890",
     *     "screen": { ... }      // Смотри структуру toECommerceScreen
     * }
     */
    static func toECommerceReferrer(referrer: [AnyHashable: Any]) -> YMMECommerceReferrer {
        var yamScreen: YMMECommerceScreen? = nil
        
        if referrer.index(forKey: "screen") != nil {
            yamScreen = toECommerceScreen(screen: referrer["screen"] as? [AnyHashable: Any] ?? [:])
        }
        
        let yamReferrer = YMMECommerceReferrer(
            type:       referrer["type"] as? String,
            identifier: referrer["identifier"] as? String,
            screen:     yamScreen
        )
        
        return yamReferrer
    }
    
    /*
     * From:
     * {
     *     "fiat": [4.53, "USD"],      // [!] Обязательный
     *     "internalComponents": [
     *          [30_570_000, "wood"],
     *          [26.89, "iron"],
     *          [5.1, "gold"]
     *     ]
     * }
     */
    static func toECommercePrice(price: [AnyHashable: Any]) throws -> YMMECommercePrice {
        let internalComponents: [YMMECommerceAmount]? = try (price["internalComponents"] as? [Any])?.map { amount in
            return try self.toECommerceAmount(amount: amount as? [Any] ?? [])
        }
        
        let actualPrice = YMMECommercePrice(
            fiat: try self.toECommerceAmount(amount: price["fiat"]! as? [Any] ?? []),
            internalComponents: internalComponents
        )
        
        return actualPrice
    }
    
    /*
     * From:
     * [10.5, "USD"]
     */
    static func toECommerceAmount(amount: [Any]) throws -> YMMECommerceAmount {
        guard
            let value = amount[0] as? NSNumber,
            let unit  = amount[1] as? String
        else {
            throw ValidationError.incorrectAmount
        }
        
        let yamAmount = YMMECommerceAmount(
            unit:  unit,
            value: NSDecimalNumber(value: value.doubleValue)
        )
        
        return yamAmount
    }
    
    static func toStringArray(_ rawItems: Any?) -> [String]? {
        guard let items = rawItems as? NSArray else {
            return nil
        }
        
        var preparedItems = [String]()

        for rawItem in items {
            // Skip no String() items
            if let strItem = rawItem as? String {
                preparedItems.append(strItem)
            }
        }

        return preparedItems
    }
    
    /*
     * From:
     * {
     *   "name": "Ivan Nalivayko",
     *   "gender": "male",
     *   "notificationEnabled": false,
     *   "birthDate": { // OR: { "age": 20 }
     *     "year": 2001,
     *     "month": 1,
     *     "day": 1
     *   }
     * }
     */
    static func toUserProfile(user: [AnyHashable: Any]) throws -> YMMMutableUserProfile {
        let yamProfile = YMMMutableUserProfile()
        
        if let name = user["name"] as? String {
            yamProfile.apply(YMMProfileAttribute.name().withValue(name))
        }
        
        if let gender = user["gender"] as? String {
            yamProfile.apply(YMMProfileAttribute.gender().withValue(toGenderType(gender)))
        }
        
        if let birthDateParts = user["birthDate"] as? [AnyHashable: Any] {
            var birthDate: YMMUserProfileUpdate? = nil;
            
            if let year = birthDateParts["year"] as? NSNumber as? UInt {
                if let month = birthDateParts["month"] as? NSNumber as? UInt {
                    if let day = birthDateParts["day"] as? NSNumber as? UInt {
                        birthDate = YMMProfileAttribute
                            .birthDate()
                            .withDate(year: year, month: month, day: day)
                    } else {
                        birthDate = YMMProfileAttribute
                            .birthDate()
                            .withDate(year: year, month: month)
                    }
                } else {
                    birthDate = YMMProfileAttribute
                        .birthDate()
                        .withDate(year: year)
                }
            } else if let age = birthDateParts["age"] as? NSNumber as? UInt {
                birthDate = YMMProfileAttribute
                    .birthDate()
                    .withAge(age)
            }
            
            if birthDate != nil {
                yamProfile.apply(birthDate!)
            }
        }
        
        if let notificationEnabled = user["notificationEnabled"] as? Bool {
            yamProfile.apply(YMMProfileAttribute.notificationsEnabled().withValue(notificationEnabled))
        }
        
        return yamProfile
    }
    
    static func toGenderType(_ gender: String) -> YMMGenderType {
        if gender == "female" {
            return YMMGenderType.female
        }
        
        if gender == "male" {
            return YMMGenderType.male
        }
        
        return YMMGenderType.other
    }
}

extension Converter.ValidationError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .apiKeyNotDefined:
            return NSLocalizedString("Api key not defined", comment: "Api key not defined")
        case .incorrectAmount:
            return NSLocalizedString("Incorrect amount", comment: "Incorrect amount value")
        case .incorrectProductSku:
            return NSLocalizedString("Incorrect SKU product", comment: "Incorrect SKU product")
        case .incorrectProductItemQty:
            return NSLocalizedString("Incorrect product qantity", comment: "Incorrect cart product item quantity")
        case .incorrectOrderId:
            return NSLocalizedString("Incorrect order id", comment: "Incorrect order id")
        case .incorrectOrderCartItems:
            return NSLocalizedString("Incorrect order cart items", comment: "Incorrect order cart items")
        case .unknown:
            return NSLocalizedString("Unknown", comment: "Unknown")
        }
    }
}
